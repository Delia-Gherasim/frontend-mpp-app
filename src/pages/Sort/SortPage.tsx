import {
    FormControl,
    InputLabel,
    MenuItem,
    Select,
    SelectChangeEvent,
} from '@mui/material';
import React, {useState} from 'react';
import SideBar from '../../component/SideBar';
import TopBar from '../../component/TopBar';
import {IDevice} from '../../model/item.type';

type Props = {
    list: IDevice[];
};

const Sort = (props: Props) => {
    const [field, setField] = React.useState<keyof IDevice | ''>('');

    const handleChange = (event: SelectChangeEvent<string>) => {
        setField(event.target.value as keyof IDevice);
    };

    const {list} = props;

    const sortedList = React.useMemo(() => {
        if (field === 'date') {
            return [...list].sort((a, b) => {
                const valueA =
                    a[field] instanceof Date ? a[field] : new Date(a[field]);
                const valueB =
                    b[field] instanceof Date ? b[field] : new Date(b[field]);
                return valueA.getTime() - valueB.getTime();
            });
        } else if (Array.isArray(list) && field) {
            return [...list].sort((a, b) => {
                const valueA = (a[field] ?? '').toString().toLowerCase();
                const valueB = (b[field] ?? '').toString().toLowerCase();
                return valueA.localeCompare(valueB);
            });
        }
        return list;
    }, [field, list]);

    const [isSidebarOpen, setIsSidebarOpen] = useState(false);
    const handleSidebarToggle = () => {
        setIsSidebarOpen(!isSidebarOpen);
    };

    return (
        <div className={`container ${isSidebarOpen ? 'sidebar-open' : ''}`}>
            <TopBar />
            <SideBar
                isOpen={isSidebarOpen}
                handleToggle={handleSidebarToggle}
            />
            <section className='section-content'>
                <div
                    style={{
                        display: 'flex',
                        justifyContent: 'center',
                        flexDirection: 'column',
                        alignItems: 'center',
                    }}
                >
                    <h2 style={{textAlign: 'center'}}>Sorted Items</h2>
                    <FormControl
                        fullWidth
                        style={{
                            borderRadius: '10px',
                            borderColor: 'black',
                            color: 'black',
                        }}
                    >
                        <InputLabel id='demo-simple-select-label'>
                            Field to sort by
                        </InputLabel>
                        <Select
                            labelId='demo-simple-select-label'
                            id='demo-simple-select'
                            value={field}
                            label='Sort By: '
                            onChange={handleChange}
                            style={{
                                borderRadius: '15px',
                                borderColor: 'black',
                                color: 'black',
                                width: '200px',
                            }}
                        >
                            <MenuItem value={'type'}>Type</MenuItem>
                            <MenuItem value={'brand'}>Brand</MenuItem>
                            <MenuItem value={'owner'}>Owner</MenuItem>
                            <MenuItem value={'date'}>Date</MenuItem>
                        </Select>
                    </FormControl>
                    <table>
                        <thead>
                            <tr>
                                <th>Type</th>
                                <th>Brand</th>
                                <th>Owner</th>
                                <th>Accessories</th>
                                <th>Warranty</th>
                                <th>Date</th>
                            </tr>
                        </thead>
                        <tbody>
                            {sortedList.map((device) => (
                                <tr key={device.id}>
                                    <td>{device.type}</td>
                                    <td>{device.brand}</td>
                                    <td>{device.owner}</td>
                                    <td>{device.accessories ? 'Yes' : 'No'}</td>
                                    <td>{device.warranty ? 'Yes' : 'No'}</td>
                                    <td>
                                        {device.date
                                            ? device.date.toDateString()
                                            : ''}
                                    </td>
                                </tr>
                            ))}
                        </tbody>
                    </table>
                </div>
            </section>
        </div>
    );
};

export default Sort;
