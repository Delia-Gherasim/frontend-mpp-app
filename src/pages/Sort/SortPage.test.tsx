import {render} from '@testing-library/react';
import {BrowserRouter as Router} from 'react-router-dom';
import {describe, expect, it} from 'vitest';
import Sort from './SortPage';

describe('Sort', () => {
    it('should render the Sort page', () => {
        const renderedComp = render(
            <Router>
                <Sort />
            </Router>,
        );
        expect(renderedComp).toMatchSnapshot();
    });
});
