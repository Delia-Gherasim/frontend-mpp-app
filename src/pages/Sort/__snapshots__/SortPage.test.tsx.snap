// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sort > should render the Sort page 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container "
      >
        <article
          class="article-header"
        >
          <header>
            <h2
              class="title"
            >
              Awesome App
            </h2>
          </header>
        </article>
        <div
          class="sidebar-container "
        >
          <div
            class="content"
          >
            <hr />
            <br />
            <a
              class="sidebar-item"
              href="/"
              style="color: rgb(230, 230, 230);"
            >
              See Database
            </a>
            <hr />
            <a
              class="sidebar-item"
              href="/search"
              style="color: rgb(230, 230, 230);"
            >
              Search
            </a>
            <hr />
            <a
              class="sidebar-item"
              href="/sort"
              style="color: rgb(230, 230, 230);"
            >
              Sort
            </a>
            <hr />
            <a
              class="sidebar-item"
              href="/chart"
              style="color: rgb(230, 230, 230);"
            >
              Chart
            </a>
            <hr />
          </div>
        </div>
        <button
          class="hide-button"
          style="background-color: rgb(218, 165, 32); border-radius: 5px; color: black;"
        >
          ☰
        </button>
        <section
          class="section-content"
        >
          <div
            style="display: flex; justify-content: center; flex-direction: column; align-items: center;"
          >
            <h2
              style="text-align: center;"
            >
              Sorted Items
            </h2>
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
              style="border-radius: 10px; border-color: black; color: black;"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                id="demo-simple-select-label"
              >
                Field to sort by
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                style="border-radius: 15px; border-color: black; color: black; width: 200px;"
              >
                <div
                  aria-controls=":r0:"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="demo-simple-select-label demo-simple-select"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="demo-simple-select"
                  role="combobox"
                  tabindex="0"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Sort By: 
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <table>
              <thead>
                <tr>
                  <th>
                    Category
                  </th>
                  <th>
                    Type
                  </th>
                  <th>
                    Brand
                  </th>
                  <th>
                    Owner
                  </th>
                  <th>
                    Accessories
                  </th>
                  <th>
                    Warranty
                  </th>
                  <th>
                    Date
                  </th>
                </tr>
              </thead>
              <tbody />
            </table>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container "
    >
      <article
        class="article-header"
      >
        <header>
          <h2
            class="title"
          >
            Awesome App
          </h2>
        </header>
      </article>
      <div
        class="sidebar-container "
      >
        <div
          class="content"
        >
          <hr />
          <br />
          <a
            class="sidebar-item"
            href="/"
            style="color: rgb(230, 230, 230);"
          >
            See Database
          </a>
          <hr />
          <a
            class="sidebar-item"
            href="/search"
            style="color: rgb(230, 230, 230);"
          >
            Search
          </a>
          <hr />
          <a
            class="sidebar-item"
            href="/sort"
            style="color: rgb(230, 230, 230);"
          >
            Sort
          </a>
          <hr />
          <a
            class="sidebar-item"
            href="/chart"
            style="color: rgb(230, 230, 230);"
          >
            Chart
          </a>
          <hr />
        </div>
      </div>
      <button
        class="hide-button"
        style="background-color: rgb(218, 165, 32); border-radius: 5px; color: black;"
      >
        ☰
      </button>
      <section
        class="section-content"
      >
        <div
          style="display: flex; justify-content: center; flex-direction: column; align-items: center;"
        >
          <h2
            style="text-align: center;"
          >
            Sorted Items
          </h2>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
            style="border-radius: 10px; border-color: black; color: black;"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              id="demo-simple-select-label"
            >
              Field to sort by
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              style="border-radius: 15px; border-color: black; color: black; width: 200px;"
            >
              <div
                aria-controls=":r0:"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="demo-simple-select-label demo-simple-select"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="demo-simple-select"
                role="combobox"
                tabindex="0"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Sort By: 
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <table>
            <thead>
              <tr>
                <th>
                  Category
                </th>
                <th>
                  Type
                </th>
                <th>
                  Brand
                </th>
                <th>
                  Owner
                </th>
                <th>
                  Accessories
                </th>
                <th>
                  Warranty
                </th>
                <th>
                  Date
                </th>
              </tr>
            </thead>
            <tbody />
          </table>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
